File: client/src/types/knowledge-object.ts
export type KnowledgeStatus =
  | "draft"
  | "pending_review"
  | "approved"
  | "rejected"
  | "archived"
  | "superseded";

export type KnowledgeRiskLevel = "low" | "medium" | "high" | "critical" | string;

export interface IKnowledgeObject {
  id: string;
  site_ids?: string[] | null;
  source_type: string;
  source_id?: string | null;
  title: string;
  topic?: string | null;
  task_type?: string | null;
  asset_name?: string | null;
  risk_level?: KnowledgeRiskLevel | null;
  problem?: string | null;
  root_cause?: string | null;
  recommended_action?: string | null;
  lesson_learned?: string | null;
  safety_warning?: string | null;
  required_ppe?: string[] | null;
  stop_work_triggers?: string[] | null;
  status: KnowledgeStatus | string;
  confidence_score?: number | string | null;
  sme_notes?: string | null;
  rejection_reason?: string | null;
  created_by?: string | null;
  approved_by?: string | null;
  approved_at?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface ICreateKnowledgeObjectPayload {
  source_type: string;
  source_id?: string;
  title: string;
  site_ids?: string[];
  topic?: string;
  task_type?: string;
  asset_name?: string;
  risk_level?: string;
  problem?: string;
  root_cause?: string;
  recommended_action?: string;
  lesson_learned?: string;
  safety_warning?: string;
  required_ppe?: string[];
  stop_work_triggers?: string[];
  confidence_score?: number;
  sme_notes?: string;
  status: KnowledgeStatus;
}

export type IUpdateKnowledgeObjectPayload = Partial<ICreateKnowledgeObjectPayload> & {
  rejection_reason?: string;
};

File: server/app/modules/knowledge/models/knowledge_object.py
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Index, Numeric, String, Text
from sqlalchemy.dialects.postgresql import ARRAY, JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.base_table import BaseTable, SoftDeleteMixin, TenantMixin


class KnowledgeObject(BaseTable, SoftDeleteMixin, TenantMixin):
    __tablename__ = "knowledge_objects"

    __table_args__ = (
        Index("ix_knowledge_objects_source_type", "source_type"),
        Index("ix_knowledge_objects_source_id", "source_id"),
        Index("ix_knowledge_objects_topic", "topic"),
        Index("ix_knowledge_objects_task_type", "task_type"),
        Index("ix_knowledge_objects_risk_level", "risk_level"),
        Index("ix_knowledge_objects_status", "status"),
        Index("ix_knowledge_objects_created_by", "created_by"),
    )

    site_ids: Mapped[list[uuid.UUID] | None] = mapped_column(
        ARRAY(UUID(as_uuid=True)),
        nullable=True,
    )
    source_type: Mapped[str] = mapped_column(String(64), nullable=False)
    source_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        nullable=True,
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    topic: Mapped[str | None] = mapped_column(String(255), nullable=True)
    task_type: Mapped[str | None] = mapped_column(String(128), nullable=True)
    asset_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    risk_level: Mapped[str | None] = mapped_column(String(32), nullable=True)
    problem: Mapped[str | None] = mapped_column(Text, nullable=True)
    root_cause: Mapped[str | None] = mapped_column(Text, nullable=True)
    recommended_action: Mapped[str | None] = mapped_column(Text, nullable=True)
    lesson_learned: Mapped[str | None] = mapped_column(Text, nullable=True)
    safety_warning: Mapped[str | None] = mapped_column(Text, nullable=True)
    required_ppe: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    stop_work_triggers: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    status: Mapped[str] = mapped_column(
        String(32),
        default="draft",
        nullable=False,
    )
    confidence_score: Mapped[Decimal | None] = mapped_column(
        Numeric(4, 3),
        nullable=True,
    )
    sme_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    rejection_reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=True,
    )
    approved_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=True,
    )
    approved_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )

    created_by_user = relationship(
        "User",
        back_populates="created_knowledge_objects",
        foreign_keys=[created_by],
    )
    approved_by_user = relationship(
        "User",
        back_populates="approved_knowledge_objects",
        foreign_keys=[approved_by],
    )
